crates/iceberg/src/expr/visitors/page_index_evaluator.rs

```diff
@@ -0,0 +1,1491 @@
+    pub(crate) fn eval(
+        filter: &'a BoundPredicate,
+        column_index: &'a [Index],
+        offset_index: &'a OffsetIndex,
+        row_group_metadata: &'a RowGroupMetaData,
+        field_id_map: &'a HashMap<i32, usize>,
+        snapshot_schema: &'a Schema,
+    ) -> Result<Vec<RowSelector>> {
+        if row_group_metadata.num_rows() == 0 {
+            return Ok(vec![]);
+        }
+
+        let mut evaluator = Self::new(
+            column_index,
+            offset_index,
+            row_group_metadata,
+            field_id_map,
+            snapshot_schema,
+        );
+
+        Ok(visit(&mut evaluator, filter)?.iter().copied().collect())
+    }
+
+    fn select_all_rows(&self) -> Result<RowSelection> {
+        Ok(vec![RowSelector::select(
+            self.row_group_metadata.num_rows() as usize
+        )]
+        .into())
+    }
+
+    fn skip_all_rows(&self) -> Result<RowSelection> {
+        Ok(vec![].into())
+    }
+
+    fn calc_row_selection<F>(
+        &self,
+        field_id: i32,
+        predicate: F,
+        missing_col_behavior: MissingColBehavior,
+    ) -> Result<RowSelection>
+    where
+        F: Fn(Option<Datum>, Option<Datum>, PageNullCount) -> Result<bool>,
+    {
+        let Some(&parquet_column_index) =
+            self.iceberg_field_id_to_parquet_column_index.get(&field_id)
+        else {
+            // if the snapshot's column is not present in the row group,
+            // exit early
+            return match missing_col_behavior {
+                MissingColBehavior::CantMatch => self.skip_all_rows(),
+                MissingColBehavior::MightMatch => self.select_all_rows(),
+            };
+        };
+
+        let Some(field) = self.snapshot_schema.field_by_id(field_id) else {
+            return Err(Error::new(
+                ErrorKind::Unexpected,
+                format!("Field with id {} missing from snapshot schema", field_id),
+            ));
+        };
+
+        let Some(field_type) = field.field_type.as_primitive_type() else {
+            return Err(Error::new(
+                ErrorKind::Unexpected,
+                format!(
+                    "Field with id {} not convertible to primitive type",
+                    field_id
+                ),
+            ));
+        };
+
+        let Some(column_index) = self.column_index.get(parquet_column_index) else {
+            // This should not happen, but we fail soft anyway so that the scan is still
+            // successful, just a bit slower
+            return self.select_all_rows();
+        };
+
+        let Some(offset_index) = self.offset_index.get(parquet_column_index) else {
+            // if we have a column index, we should always have an offset index.
+            return Err(Error::new(
+                ErrorKind::Unexpected,
+                format!("Missing offset index for field id {}", field_id),
+            ));
+        };
+
+        let row_counts = self.calc_row_counts(offset_index);
```

Comment:
Good spot. I think the optimal approach here would not be to calc row counts for all columns up-front because only the ones for columns that appear in the predicate would be needed. Caching the result would make more sense. If it is ok with you I'll add a TODO for this rather than address in this PR, and we can address in a follow-up, as it is not broken and caching would not save us a huge amount of time here.